enhancements/olm/recursive-cluster-operators.md

```diff
@@ -0,0 +1,415 @@
+The CVO manages [hundreds of in-cluster resources via manifests](/dev-guide/cluster-version-operator/user/reconciliation.md#manifest-graph), including tens of "second level operators".
+Management includes:
+
+* Monitoring operand status, so we can complain to the admin (via ClusterVersion and alerts) if an operand is having difficulty.
+* Passing configuration changes (e.g. admin requests an update via ClusterVersion `spec.desiredUpdate.version="4.9.99"`, etc.), on to operands (e.g. setting `spec.containers[].image` in the etcd operator Deployment).
+* Managing interactions between operands (e.g. waiting for the Kubernetes API operator to report completed updates before asking dependent components like the machine-config operator to update).
+
+This is a generic operator pattern, and not something that is unique to the CVO.
+
+In many situations, operand resource types have a `status` that reports on their direct status.
+For example, the Deployment controller sets [`status` properties on managed Deployments][Deployment-status], like "`Progressing=True` with `reason=NewReplicaSetAvailable`; we are [complete][Deployment-complete]".
+That includes information about underlying ReplicaSets, which the Deployment controller manages on behalf of the Deployment, and recursively from Pods, which the ReplicaSet controller manages on behalf of the ReplicaSets.
+But it does not include information from an operand branching out to manage something directly, like the etcd operator managing etcd members.
+The Deployment controller has no idea what's going on with etcd members.
+
+For all of OpenShift v4, we have had a a [ClusterOperator type][ClusterOperator].
+This gives operators who manage in-cluster resources a pathway in parallel with their own Deployment to reach around the Deployment controller and pass information directly back to the CVO.
+For example, the etcd operator can say "etcd members and my other operands are healthy, and we're currently all at 4.9.99".
+Between the etcd ClusterOperator `status` and Deployment `status`, the CVO has a complete picture of how the etcd operator is doing.
+
+The CVO does not have a monopoly on having operators as its operands.
+For example, OLM is is the business of installing and managing operators.
+Something installs and manages the Deployment operator.
+The monitoring operator installs and manages a Prometheus operator.
+All of these operators need a way to get information back out of their operands about how their operating is going, so they can in turn pass that information up to their parents.
+The ClusterOperator type is a very generic `status` object, and would be a good vessel for transmitting this information.
+
+### Tracking ownership
+
+No need for a `spec.parent` property or anything; we already have `metadata.ownerReferences`.
+Might need a namespaced variant?  I dunno how many operators can actually be installed with the same name in parallel namespaces.
```

Comment:
Assume identical operator installs in multiple namespaces.  For example, ROSA/OSD addons today provision their own monitoring stack using prometheus operator (not UWM).  I think they use the same pattern.  So if 2+ addons are installed that need prometheus you have prometheus in 2+ namespaces.

Reply:
Yeah, could do with with two `Operator` (or whatever you call the namespaced variant) in different namespaces that both use the `prometheus` name, or with two `ClusterOperator` that use different names `openshift-prometheus` and `openshift-user-workload-prometheus`, or whatever).